chainerx_cc/chainerx/cuda/cuda_device/linalg.cu

```diff
@@ -0,0 +1,208 @@
+}
+
+template <>
+cusolverStatus_t GetrfBuffersize<double>(cusolverDnHandle_t handle, int m, int n, double* a, int lda, int* lwork) {
+    return cusolverDnDgetrf_bufferSize(handle, m, n, a, lda, lwork);
+}
+
+template <>
+cusolverStatus_t GetrfBuffersize<float>(cusolverDnHandle_t handle, int m, int n, float* a, int lda, int* lwork) {
+    return cusolverDnSgetrf_bufferSize(handle, m, n, a, lda, lwork);
+}
+
+template <>
+cusolverStatus_t Getrf<double>(cusolverDnHandle_t handle, int m, int n, double* a, int lda, double* workspace, int* devipiv, int* devinfo) {
+    return cusolverDnDgetrf(handle, m, n, a, lda, workspace, devipiv, devinfo);
+}
+
+template <>
+cusolverStatus_t Getrf<float>(cusolverDnHandle_t handle, int m, int n, float* a, int lda, float* workspace, int* devipiv, int* devinfo) {
+    return cusolverDnSgetrf(handle, m, n, a, lda, workspace, devipiv, devinfo);
+}
+
+template <>
+cusolverStatus_t Getrs<double>(
+        cusolverDnHandle_t handle,
+        cublasOperation_t trans,
+        int n,
+        int nrhs,
+        double* a,
+        int lda,
+        int* devipiv,
+        double* b,
+        int ldb,
+        int* devinfo) {
+    return cusolverDnDgetrs(handle, trans, n, nrhs, a, lda, devipiv, b, ldb, devinfo);
+}
+
+template <>
+cusolverStatus_t Getrs<float>(
+        cusolverDnHandle_t handle,
+        cublasOperation_t trans,
+        int n,
+        int nrhs,
+        float* a,
+        int lda,
+        int* devipiv,
+        float* b,
+        int ldb,
+        int* devinfo) {
+    return cusolverDnSgetrs(handle, trans, n, nrhs, a, lda, devipiv, b, ldb, devinfo);
+}
+
+template <typename T>
+void SolveImpl(const Array& a, const Array& b, const Array& out) {
+    Device& device = a.device();
+    Dtype dtype = a.dtype();
+
+    cuda_internal::DeviceInternals& device_internals = cuda_internal::GetDeviceInternals(static_cast<CudaDevice&>(device));
+
+    Array lu_matrix = Empty(a.shape(), dtype, device);
+    device.backend().CallKernel<CopyKernel>(a.Transpose(), lu_matrix);
+    T* lu_ptr = static_cast<T*>(internal::GetRawOffsetData(lu_matrix));
+
+    int m = a.shape()[0];
+    int nrhs = 1;
+    if (b.ndim() == 2) {
+        nrhs = b.shape()[1];
+    }
+
+    Array ipiv = Empty(Shape{m}, Dtype::kInt32, device);
+    int* ipiv_ptr = static_cast<int*>(internal::GetRawOffsetData(ipiv));
+
+    int buffersize = 0;
+    device_internals.cusolverdn_handle().Call(GetrfBuffersize<T>, m, m, lu_ptr, m, &buffersize);
+
+    Array work = Empty(Shape{buffersize}, dtype, device);
+    T* work_ptr = static_cast<T*>(internal::GetRawOffsetData(work));
+
+    std::shared_ptr<void> devinfo = device.Allocate(sizeof(int));
+
+    device_internals.cusolverdn_handle().Call(Getrf<T>, m, m, lu_ptr, m, work_ptr, ipiv_ptr, static_cast<int*>(devinfo.get()));
+
+    int devinfo_h = 0;
+    Device& native_device = dynamic_cast<native::NativeDevice&>(GetDefaultContext().GetDevice({"native", 0}));
```

Comment:
Do you need the dynamic cast even when it's assigned to a Device&?

Reply:
It works without dynamic cast, it's not needed. Borrowed the code from [here](https://github.com/chainer/chainer/blob/master/chainerx_cc/chainerx/native/native_device_test.cc#L29).

Reply:
Let's remove the cast then. The reason it's needed https://github.com/chainer/chainer/blob/master/chainerx_cc/chainerx/native/native_device_test.cc#L29 is because the return type is a NativeDevice& and not a Device&.